iotilecore/iotile/core/hw/debug/firmware_image_analyzer.py

```diff
@@ -0,0 +1,140 @@
+from iotile.core.utilities.intelhex import IntelHex
+from iotile.core.exceptions import *
+from tempfile import NamedTemporaryFile
+import os
+import subprocess
+import struct
+
+class FirmwareImageAnalyzer:
+    """A firmware image analyzer using IntelHex"""
+
+    SIZE_OF_CDBAPPINFO = 32
+    MAGIC_NUMBER_BYTE_OFFSET = 24
+
+    def __init__(self, firmware, magic_number=0xBAADDAAD):
+        if not firmware.endswith(".elf") and not firmware.endswith(".hex"):
+            raise ArgumentError("You must pass an ARM firmware image in elf/hex format", path=firmware)
+
+        tmpf = NamedTemporaryFile(delete=False)
+        tmpf.close()
+
+        tmp = tmpf.name
+
+        try:
+            if firmware.endswith(".elf"):
+                err = subprocess.call(['arm-none-eabi-objcopy', '-O', 'ihex', firmware, tmp])
+                if err != 0:
+                    raise ExternalError("Cannot convert elf to binary file", error_code=err)
+
+                hex_image = IntelHex(tmp)
+            else:
+                hex_image = IntelHex(firmware)
+
+            self._hex_image = hex_image
+        finally:
+            os.remove(tmp)
+
+        self.magic_number = magic_number
+        self.min_addr = self._hex_image.minaddr()
+        self.max_addr = self._hex_image.maxaddr()
+        self.num_addresses = len(self._hex_image.addresses())
+        self.segments = self._hex_image.segments()
+        self.memory_size = self._hex_image.get_memory_size()
+        self.cdb_app_info = self.get_cdb_app_info()
+        if self.cdb_app_info is None:
+            raise DataError("Unable to analyze firmware image.")
+
+    def get_firmware_hardware_type(self):
+        """Returns hardware type specified in cdb app info"""
+        if self.cdb_app_info is None:
+            self.get_cdb_app_info()
+
+        return self.cdb_app_info['hardware_type']
+
+    def get_firmware_api_version(self):
+        """Returns firmware api version specified in cdb app info"""
+        if self.cdb_app_info is None:
```

Comment:
You can get rid of all of these checks since you know if the object was created you have a `cdb_app_info`.  That's one of the key benefits of doing processing during `__init__`.  A user doesn't get an instance unless the __init__ processing was successful.

Reply:
You right. Fixed.